rero_ils/modules/patron_transaction_events/serializers/csv.py

```diff
@@ -0,0 +1,190 @@
+                if pid := event.get('library', {}).get('pid'):
+                    record = self.get_resource(Library, pid)
+                    csv_data |= _extract_transaction_library_data(record)
+                    csv_data['transaction_date'] = transaction_date\
+                        .astimezone(tz=record.get_timezone())\
+                        .isoformat()
+                if pid := parent.loan_pid:
+                    loan = self.get_resource(Loan, pid)
+                    document = self.get_resource(Document, loan.document_pid)
+                    item = self.get_resource(Item, loan.item_pid)
+                    csv_data |= _extract_document_data(document)
+                    csv_data |= _extract_item_data(item, self)
+
+                # write csv data
+                data = self.process_dict(csv_data)
+                writer.writerow(data)
+                yield line.read()
+
+        # return streamed content
+        return stream_with_context(generate_csv())
+
+
+def _extract_patron_data(record):
+    """Extract data from patron to include into the csv export file.
+
+    :param record: the `Patron` to analyze.
+    :returns a dictionary containing desired patron data.
+    """
+    data = {
+        'patron_name':
+            record.formatted_name,
+        'patron_barcode':
+            ', '.join(record.get('patron', {}).get('barcode', [])),
+        'patron_email':
+            record.user.email or
+            record.get('patron', {}).get('additional_communication_email')
+    }
+    return {k: v for k, v in data.items() if v}
+
+
+def _extract_document_data(record):
+    """Extract document from loan to include into the csv export file.
+
+    :param record: the `Document` representing the document to analyze.
+    :returns a dictionary containing desired document data.
+    """
+    data = {
+        'document_pid': record.pid,
+        'document_title':
+            title_format_text_head(record.get('title', []))
+    }
+    return {k: v for k, v in data.items() if v}
+
+
+def _extract_item_data(record, collector):
+    """Extract item from loan to include into the csv export file.
+
+    :param record: the `Item` representing the item to analyze.
+    :param collector: the class that call this method.
+    :returns a dictionary containing desired item data.
+    """
+    library = collector.get_resource(Library, record.library_pid)
+    data = {
+        'item_pid': record.pid,
+        'item_barcode': record.get('barcode'),
+        'item_owning_library': library.get('name')
+    }
+    return {k: v for k, v in data.items() if v}
+
+
+def _extract_operator_data(record):
+    """Extract data from operator to include into the csv export file.
+
+    :param record: the `Patron` representing the operator to analyze.
+    :returns a dictionary containing desired operator data.
+    """
+    data = {
+        'operator_name': record.formatted_name
+    }
+    return {k: v for k, v in data.items() if v}
```

Comment:
dict comprehension seem useless

Reply:
for all these dict_comprenhension : the purpose is the remove empty entry if value is None.
I won't know/ensure than a method return always a value. So I build a dict and clean this dict.
So If one day, a called method/property (formatted_name, record.getNumberOfItems()) are updated and return None, my serializer will continue to work fine.